### PR TITLE
[WIP] Migrate to Flask-SQLAlchemy 3.1.1 and SQLAlchemy 2.0.23

### PR DESCRIPTION
**Pull request by DjVinnii**
## What type of PR?

Enhancement

## What does this PR do?
Upgrade to Flask-SQLAlchemy 3.1.1 and SQLAlchemy 2.0.23.

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [N/A] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

**Comment by DjVinnii** [comment]
I think I tested most of this PR, but it is possible I forgot something. Maybe it is good to improve local testing experience with proper unit/feature tests for at least the admin part.  For now, it would be great if someone can take a look for initial review.

**Comment by nextgens** [comment]
You are more than welcome to add tests :)

bors try

**Comment by Diman0** [comment]
bors try

File: core/admin/mailu/manage.py [hunk 1]
```diff
@@ -96,7 +97,7 @@ def admin(localpart, domain_name, password, mode):
 def user(localpart, domain_name, password):
     """ Create a user
     """
-    domain = models.Domain.query.get(domain_name)
+    domain = db.session.get(models.domain, domain_name)
```
Comment (Diman0) [inline core/admin/mailu/manage.py:100]: This fails during the automatic test with the error:
```
   File "/app/mailu/manage.py", line 100, in user
    domain = db.session.get(models.domain, domain_name)
                            ^^^^^^^^^^^^^
AttributeError: module 'mailu.models' has no attribute 'domain'. Did you mean: 'Domain'?
```

full stack trace at: https://github.com/Mailu/Mailu/actions/runs/6868358963/job/18678819073#step:11:1197

**Review by Diman0 — CHANGES_REQUESTED** [review]
Tests do not pass. See in-line comment

**Comment by DjVinnii** [comment]
bors try

File: core/admin/mailu/manage.py (hunk 1 above)
Comment (DjVinnii) [reply core/admin/mailu/manage.py]: This should be `models.Domain` indeed, fixed.

**Comment by nextgens** [comment]
bors retry

**Comment by nextgens** [comment]
bors try

File: core/admin/mailu/internal/views/dovecot.py [hunk 2]
```diff
@@ -20,16 +21,18 @@ def dovecot_passdb_dict(user_email):
...
 
+
 @internal.route("/dovecot/userdb/<path:user_email>")
 def dovecot_userdb_dict(user_email):
     try:
-        quota = models.User.query.filter(models.User.email==user_email).with_entities(models.User.quota_bytes).one_or_none() or flask.abort(404)
+        quota = db.one_or_404(db.select(models.User.quota_bytes).filter_by(email=user_email))
     except sqlalchemy.exc.StatementError as exc:
         flask.abort(404)
     return flask.jsonify({
```
Comment (Diman0) [inline core/admin/mailu/internal/views/dovecot.py:38]: Latest CI error is for this function. It is used when someone logs in via imap. 


https://github.com/Mailu/Mailu/actions/runs/6870165755/job/18684540112#step:11:1755
```
 [2023-11-14 22:19:48,356] INFO in nginx: Login attempt for: admin@mailu.io/submission/587 from: 192.168.203.1/36240: success: password
[2023-11-14 22:19:49,176] ERROR in app: Exception on /internal/dovecot/userdb/user@mailu.io [GET]
Traceback (most recent call last):
  File "/app/venv/lib/python3.11/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.11/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.11/site-packages/flask_restx/api.py", line 674, in error_router
    return original_handler(e)
           ^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.11/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.11/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/mailu/internal/views/dovecot.py", line 39, in dovecot_userdb_dict
    "quota_rule": f"*:bytes={quota[0]}"
                             ~~~~~^^^
TypeError: 'int' object is not subscriptable
```

**Review by Diman0 — CHANGES_REQUESTED** [review]
CI caught another error. See in-line comment. I'm not sure what the problem is. This function should be called when you access imap (dovecot). Dovecot should use this to lookup the user.

File: core/admin/mailu/internal/views/dovecot.py (hunk 2 above)
Comment (nextgens) [reply core/admin/mailu/internal/views/dovecot.py:38]: That query is a relatively "hot path", we should keep the .with_entities() equivalent to (a) reduce the size of what is returned and (b) make it easier on the DBM.

@DjVinnii please ensure you "scope" the selects everywhere we do in the existing code.

Comment (DjVinnii) [reply core/admin/mailu/internal/views/dovecot.py:38]: This is my mistake, I assumed the  `db.select(models.User.quota_bytes)` wouldn't only return the quota_bytes but an object.

According to the SQLAlchemy documentation the `.with_entities()` is part of the legacy query API and replaces the SELECT with the given entities. I think there are 2 way to tackle this:
1. Only select the `quota_bytes` and not the whole model so there is no need to use an index to select the quota bytes.
2. Select the whole model and use `.with_only_columns()` and use the index to target the quota bytes.

For now I went with option 2 in order to keep the changes as close as to the original. I also did the same for other queries with `.with_entities()`

**Comment by nextgens** [comment]
bors retry

**Comment by nextgens** [comment]
#3138 is just a lot less invasive... so let's go for that instead

**Comment by DjVinnii** [comment]
@nextgens I fully agree with you. Unfortunately I haven't had the time to fully complete this PR. I still think it is good to move to [the 2.0 style ORM queries](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-orm-usage). If you think so as well, I'd still like to work on this (probably in smaller PR's so it is more manageable). What do you think?

**Pull request closed without merging**